terraform_templates/iam_template.py

```diff
@@ -10,10 +15,107 @@ def terraform_iam_template(config: Mapping):
+            [__iam_policy_resource(config, account["account_name"]) for account in config["aws"]["accounts"]] +
+            [__iam_role_policy_attachment(config, account["account_name"]) for account in config["aws"]["accounts"]]
+    }
+
+    return dict_template
+
+
+# Creates the IAM role that will be deployed
+def __iam_role_resource(config: Mapping, account_name: str) -> Mapping:
+    dict_template = {
+        "aws_iam_role": {
+            config["aws"]["IAM_role_name"] + "_" + account_name: {
+                "provider": "aws." + account_name,
+                "name": config["aws"]["IAM_role_name"],
+                "tags": {
+                    "owner": config["aws"]["resource_owner"]
+                },
+                "description": "The role that CloudCustodian will be acting as to implement cloud enforcement policies. "
+                               "See https://github.com/ucsc-cgp/cgp-cloud-policies.",
+                "assume_role_policy": json.dumps({
+                    "Version": "2012-10-17",
+                    "Statement": [
+                        {
+                            "Effect": "Allow",
+                            "Principal": {
+                                "AWS": config["aws"]["provider"]["principal"]
+                            },
+                            "Action": "sts:AssumeRole",
+                            "Condition": {}
+                        },
+                        {  # we need to allow AWS Lambda to use this role, as this is how the custodian policies will take actions
+                            "Effect": "Allow",
+                            "Principal": {
+                                "Service": "lambda.amazonaws.com"
+                            },
+                            "Action": "sts:AssumeRole",
+                            "Condition": {}
+                        }
+                    ]
+                })
+            }
+        }
+    }
+
+    return dict_template
+
+
+# Creates the IAM policy that will give the role permissions
+def __iam_policy_resource(config: Mapping, account_name: str) -> Mapping:
+    dict_template = {
+        "aws_iam_policy": {
+            config["aws"]["IAM_policy_name"] + "_" + account_name: {
+                "provider": "aws." + account_name,
+                "name": config["aws"]["IAM_policy_name"],
+                "description": "The policy that will give the IAM role permissions to deploy custodian resources. "
+                               "See https://github.com/ucsc-cgp/cgp-cloud-policies.",
+                "policy": json.dumps({
+                    "Version": "2012-10-17",
+                    "Statement": [
+                        {
+                            "Action": [
+                                "*"
+                            ],
+                            "Effect": "Allow",
+                            "Resource": "*"
```

Comment:
This seems pretty broad (allow everything on everything)... what is this needed for exactly?  I would try to limit this to only the use-cases it will be used for.

Reply:
Tuned down the permissions. Now, the majority of the actions are required for the default deployment of CloudCustodian, and added some specific permissions to query information about s3 buckets.